### PR TITLE
Implement Krawczyk method

### PR DESCRIPTION
**Pull request by Kolaru**
- Implement new contractor type `Krawczyk` and Krawczyk operator `𝒦`
- Implement related tests
- Clean up and uniformized tests

### CODE REVIEW

File: src/contractors.jl [hunk 1]
```diff
@@ -72,20 +69,69 @@ function 𝒩{T}(f, f′, X::Interval{T})
...
+@generated function 𝒦(f, jacobian, X::IntervalBox{N, T}) where {N, T}
+    unit = eye(N)
+
+    ex = quote
+        m = mid(X, where_bisect)
+        J = jacobian(X)
+        Y = inv(jacobian(m))
+        m = IntervalBox(Interval.(m))
+
+        IntervalBox(m - Y*f(m) + ($unit - Y*J)*(X - m))
```
Comment (dpsanders) [inline src/contractors.jl]: You should just be able to use `I` here and get rid of the `@generated`.

File: src/roots.jl [hunk 2]
```diff
@@ -78,6 +78,7 @@ contains_zero(X::IntervalBox) = all(contains_zero.(X))
 
 
 IntervalLike{T} = Union{Interval{T}, IntervalBox{T}}
+DerivativeType = Union{Type{Newton}, Type{Krawczyk}}
```
Comment (dpsanders) [inline src/roots.jl]: Make call this `NewtonLike`?

**Comment by dpsanders** [comment]
Looks good, thanks a lot!

File: src/contractors.jl [hunk 3]
```diff
@@ -72,20 +69,69 @@ function 𝒩{T}(f, f′, X::Interval{T})
...
 end
 
 
+# Krawczyk
+struct Krawczyk{F, FP} <: Contractor{F}
+    f::F
+    f′::FP   # use \prime<TAB> for ′
+end
+
+function (C::Krawczyk)(X, tol)
```
Comment (dpsanders) [inline src/contractors.jl]: This function is (almost?) exactly the same as the equivalent for Newton, which is why my idea had been to combine the two into a single type. It's not necessary to use a single type, but I think this function and the equivalent for Newton should be able to be combined.

Comment (dpsanders) [reply src/contractors.jl]: E.g. define a function `operator(C::Krawczyk) = 𝒦`.
I *think* this should be efficient.

File: src/contractors.jl [hunk 4]
```diff
@@ -72,20 +69,69 @@ function 𝒩{T}(f, f′, X::Interval{T})
...
+        J = jacobian(X)
+        Y = inv(jacobian(m))
+        m = IntervalBox(Interval.(m))
+
+        IntervalBox(m - Y*f(m) + ($unit - Y*J)*(X - m))
+    end
+    ex
+end
+
+IntervalArithmetic.mid(X::IntervalBox, α) = mid.(X, α)
```
Comment (dpsanders) [inline src/contractors.jl]: I think we can get rid of this.

File: src/contractors.jl [hunk 5]
```diff
@@ -72,20 +69,69 @@ function 𝒩{T}(f, f′, X::Interval{T})
...
+
+    m - Y*f(m) + (1 - Y*f′(X))*(X - m)
+end
+
+
+@generated function 𝒦(f, jacobian, X::IntervalBox{N, T}) where {N, T}
+    unit = eye(N)
+
+    ex = quote
+        m = mid(X, where_bisect)
```
Comment (dpsanders) [inline src/contractors.jl]: I think we can get rid of `where_bisect`.

File: src/contractors.jl [hunk 6]
```diff
@@ -72,20 +69,69 @@ function 𝒩{T}(f, f′, X::Interval{T})
...
 
 
-IntervalArithmetic.mid(X::IntervalBox, α) = mid.(X, α)
-
 doc"""
 Multi-variable Newton operator.
 """
 function 𝒩(f::Function, jacobian::Function, X::IntervalBox)  # multidimensional Newton operator
-
     m = IntervalBox(Interval.(mid(X, where_bisect)))
```
Comment (dpsanders) [inline src/contractors.jl]: I think we can get rid of `where_bisect`, since calling `mid` without the second argument now does this automatically.

Comment (dpsanders) [reply src/contractors.jl]: Ah yes, you're totally right, sorry!

Comment (dpsanders) [reply src/contractors.jl]: Ouch that division is nasty...

Comment (dpsanders) [reply src/contractors.jl]: Not sure what that should give.

File: test/roots.jl [hunk 7]
```diff
@@ -66,32 +102,43 @@ end
     XX = IntervalBox(-5..6, 2)
     tol = 1e-5
 
+    # Newton
     rts = roots(gradf, XX, Newton, tol)
     @test length(rts) == 25
+    @test all_unique(rts)
     @test rts == roots(gradf, XX, Newton, tol; deriv = xx -> ForwardDiff.jacobian(gradf, xx))
+
+    # Krawczyk
```
Comment (dpsanders) [inline test/roots.jl]: This part and Newton above can be done with a `for` loop instead of duplicating the code.

File: src/contractors.jl (hunk 6 above)
Comment (Kolaru) [reply src/contractors.jl]: In fact we have

```julia
0/X == [0, 0]
```

for any `X` that contains 0.

A quick look at the IEEE standard did not help much (I may read that in full someday to really understand all the stuff about flavors and decorated intervals). Note that `extended_div` suffers the same problem for now

```julia
julia> extended_div(0..0, -1..1)
([0, 0], ∅)
```

but here the standard  definition (10.5.5 Two output division) indicates that it should return `([-∞, ∞], ∅)`.

Comment (dpsanders) [reply src/contractors.jl]: Yes, I agree that `extended_div` is wrong here, according to the standard and comparing with the Octave interval arithmetic package; cc @eeshan9815 .

On the other hand, for standard division the situation is not so clear; the Octave package also returns `[0, 0]` in this case.

File: src/contractors.jl [hunk 8]
```diff
@@ -23,38 +29,54 @@ function (contractor::Bisection)(X, tol)
...
 
-Base.isinf(X::IntervalBox) = any(isinf.(X))
+    Contraction operation for contractors using the first derivative of the
+    function. This contraction use a bisection scheme to refine the intervals
+    with `:unkown` status.
 
-function (C::Newton)(X, tol)
+    Currently `Newton` and `Krawczyk` contractors uses this.
+"""
+function newtonlike_contract(op, C, X, tol)
```
Comment (Kolaru) [inline src/contractors.jl]: I am not sure this solution is ideal. However I feel like the most natural way of doing this would be to not have the contractors callable, but to have a method

```julia
contract(C::Contractor, X, tol)
```

and to use multiple dispatch/generated functions to handle everything correctly. Since it would be a deeper design change I wanted to discuss it first and it may be better in a separate PR.

Comment (dpsanders) [reply src/contractors.jl]: That definitely sounds like an interesting idea. I'm not sure exactly what it would gain - I guess that's why a PR would be useful.

**Comment by dpsanders** [comment]
Please don't include merge commits. I think you should be rebasing instead:

```
git checkout master
git pull upstream master

git checkout <your-branch>
git rebase master
```

Here, `upstream` is the remote that points to this repo.

**Comment by dpsanders** [comment]
Ah and then `git push -f` to force push the changes to the current branch.

File: src/contractors.jl (hunk 8 above)
Comment (dpsanders) [reply src/contractors.jl]: I feel that it's more natural to call a contractor `C` as `C(X)` to reflect the mathematical structure. This is just as possible to dispatch on, no? Or what do you gain by introducing the `contract` function?

**Comment by dpsanders** [comment]
Looks good to me, thanks a lot!

File: src/contractors.jl (hunk 8 above)
Comment (dpsanders) [reply src/contractors.jl]: Thanks, yeah I was trying to do what you're suggesting and ran into the same frustrating problem :/ 
Let's leave it like this for the moment.
